vochain/account.go

```diff
@@ -192,94 +171,406 @@ func (v *State) VerifyAccountBalance(message, signature []byte, amount uint64) (
 	return acc.Balance >= amount, address, nil
 }
 
-func (v *State) SetAccountInfoURI(address common.Address, infoURI string) error {
-	if infoURI == "" {
-		return fmt.Errorf("cannot set an empty URI")
+// AccountFromSignature extracts an address from a signed message and returns an account if exists
+func (v *State) AccountFromSignature(message, signature []byte) (*common.Address, *Account, error) {
+	pubKey, err := ethereum.PubKeyFromSignature(message, signature)
+	if err != nil {
+		return &common.Address{}, nil, fmt.Errorf("cannot extract public key from signature: %w", err)
 	}
-	var acc Account
-	v.Tx.Lock()
-	defer v.Tx.Unlock()
-	raw, err := v.Tx.DeepGet(address.Bytes(), AccountsCfg)
-	if err != nil && !errors.Is(err, arbo.ErrKeyNotFound) {
-		return err
-	} else if err == nil {
-		if err := acc.Unmarshal(raw); err != nil {
-			return err
-		}
+	address, err := ethereum.AddrFromPublicKey(pubKey)
+	if err != nil {
+		return &common.Address{}, nil, fmt.Errorf("cannot extract address from public key: %w", err)
 	}
-	acc.InfoURI = infoURI
-	accBytes, err := acc.Marshal()
+	if address == types.EthereumZeroAddressBytes {
+		return &common.Address{}, nil, fmt.Errorf("invalid address")
+	}
+	acc, err := v.GetAccount(address, false)
+	if err != nil {
+		return &common.Address{}, nil, fmt.Errorf("cannot get account: %v", err)
+	}
+	if acc == nil {
+		return &common.Address{}, nil, ErrAccountNotExist
+	}
+	return &address, acc, nil
+}
+
+// SetAccountInfoURI sets a given account infoURI
+func (v *State) SetAccountInfoURI(accountAddress common.Address, infoURI string) error {
+	acc, err := v.GetAccount(accountAddress, false)
 	if err != nil {
 		return err
 	}
-	return v.Tx.DeepSet(address.Bytes(), accBytes, AccountsCfg)
+	if acc == nil {
+		return ErrAccountNotExist
+	}
+	acc.InfoURI = infoURI
+	return v.SetAccount(accountAddress, acc)
 }
 
-func (v *State) CreateAccount(address common.Address, infoURI string, delegates []common.Address, initBalance uint64) error {
+// Create account creates an account
+func (v *State) CreateAccount(accountAddress common.Address,
+	infoURI string,
+	delegates []common.Address,
+	initBalance uint64,
+) error {
 	// check valid address
-	if address.String() == types.EthereumZeroAddress {
+	if bytes.Equal(accountAddress.Bytes(), types.EthereumZeroAddressBytes[:]) {
```

Comment:
FYI, you can replace all instances of `bytes.Equal(foo.Bytes(), types.EthereumZeroAddressBytes[:])` with just `foo == types.EthereumZeroAddressBytes` - you can compare array values directly.